### ISSUE TITLE
Nested Complex type of Nested Complex type is not updated with Entity

### ISSUE THREAD

**Issue by SergeyGolodyuk** [aspnet/EntityFrameworkCore#14713]
Sorry for confusing title but I am not sure how describe the issue in short.

Recently we decided to make our domain objects more object oriented and first step was to introduce Value Objects for some information like "Phone", "Address", "Person". In one case we got the situation where Domain entity "Account" had nested object "Person" which also had nested object "Phone". We were able to configure such model with EF Fluent API and at first glance everything worked fine but then we found little issue. While updating "Person" object of the "Account" entity (assigning whole new Person object) and saving changes to database, "Phone"object is "rolled back" to old values.

Here is minimal example project to reproduce the issue:

```c#
using Microsoft.EntityFrameworkCore;
using System;
using System.Linq;

namespace EfCoreBugReproduce
{
    class Program
    {
        static void Main(string[] args)
        {
            var opt = new DbContextOptionsBuilder<Context>()
             .UseSqlServer("Data Source=(local);Initial Catalog=ef-bug-reproduce;Integrated Security=true;");

            var ctx = new Context(opt.Options);

            var account = ctx.Accounts.FirstOrDefault();
            Console.WriteLine(account.Contact.Phone.Number);
            account.Contact = new Person { Name = "NewName", Phone = new Phone { Number = "<new num>" } };
            ctx.Update(account);

            ctx.SaveChanges();
            Console.WriteLine(account.Contact.Phone.Number);
        }
    }

    class Account
    {
        public Guid Id { get; set; }
        public Person Contact { get; set; }
    }

    class Person
    {
        public string Name { get; set; }
        public Phone Phone { get; set; }
    }

    class Phone { public string Number { get; set; } }

    class Context : DbContext
    {
        public Context(DbContextOptions options) : base(options) { }

        public DbSet<Account> Accounts { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            var accountBuilder = modelBuilder.Entity<Account>();
            accountBuilder.HasKey(a => a.Id);

            var contact = accountBuilder.OwnsOne(a => a.Contact);
            contact.Property(p => p.Name).HasColumnName("Name").HasMaxLength(100);

            var contactPhone = contact.OwnsOne(p => p.Phone);
            contactPhone.Property(p => p.Number).HasColumnName("Number").HasMaxLength(11);
        }
    }
}
```
And here is script to create database table if you need one:
```SQL
CREATE TABLE [dbo].[Accounts](
	[Id] [uniqueidentifier] NOT NULL,
	[Name] [nvarchar](100) NULL,
	[Number] [nvarchar](11) NULL,
 CONSTRAINT [PK_Accounts] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO
INSERT [dbo].[Accounts] ([Id], [Name], [Number]) VALUES (newid(), N'Some Name', N'<some num>')
```

So in the "Main" method we read an "Account" object from database then update "Contact" property by assigning a new object to it and save changes to database. After all that "Phone.Number" property will be still the same that we read from database and not the one we assigned in new "Person" object. As a temporary workaround for the issue we added extra line of code before assigning new "Person" object to account contact. It is needed to update "Phone" property of the old "Person" object and then assign a new one, something like this:
```c#
...
account.Contact.Phone = new Phone { Number = "<new num>" };
account.Contact = new Person { Name = "NewName", Phone = new Phone { Number = "<new num>" } };
...
```

Please, let us know if this is expected behavior or it will be fixed in future.

### Further technical details
EF Core version: 2.2.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 8.1
IDE: Visual Studio 2017 15.7.4

**Comment by madoxdev** [comment]
I can confirm similar issue. 
Some details:

`public class Order {
          private Order()
          {
              //some code to initialise object
          }
          //more properties
          public Prepayment Prepayment { get; private set; }
          //more methods and properties
        }`

` public class Prepayment:ValueObject<Prepayment>
    {
        private Prepayment()
        {   
        }

        public Money AmountPrepaid { get; private set; }
        public bool HasPrepaymentBeenTaken => AmountPrepaid.Amount > 0;
}`

`public class Money {
        private Money()
        {
        }
        private Money(decimal amount)
        : this()
        {
            Amount = amount;
        }
        public decimal Amount { get; private set; }
}`

`modelBuilder.Entity<Order>()
                .OwnsOne(x => x.Prepayment,
                    prepayment =>
                    {
                        prepayment.OwnsOne(x => x.AmountPrepaid,
                            amountPrepaid =>
                            {
                                amountPrepaid.Property(x => x.Amount)
                                    .HasColumnName("PrepaymentAmount")
                                    .HasColumnType("decimal(7,2)");
                            });
                    });`

`public async Task<int> SaveAsync(Order order)
        {
            if (order.Id == 0)
            {
                await _context.AddAsync(order);
            }
         
            return await _context.SaveChangesAsync();
        }`

`public async Task<int> Handle(EditOrderCommand request, CancellationToken cancellationToken)
        {
            var order = await _orderRepository.GetOrderByIdAsync(request.Id);

            var customer = new Customer(
                request.FirstName,
                request.LastName,
                request.TelephoneNumber);

            var prepayment = new Prepayment(
                Money.SomeMoney(
                    request.PrepaymentAmount
                    )
                );

            order.ApplyChanges(
                    request.UserId, 
                    request.AdditionalInformation,
                    collectionDate,
                    customer,
                    prepayment);

            await _orderRepository.SaveAsync(order);

            return order.Id;
        }`

INSERT working as expected and nested value object is updated properly. Update isn't working as expected though. It looks like the issue is presenting itself when nesting comes to play. The Order class from my example has other Value Objects as well, and those works as expected.

**Comment by AndriySvyryd** [comment]
This was fixed in 3.0
